src/nomad_simulations/schema_packages/data_types.py

```diff
@@ -0,0 +1,260 @@
+#
+# Copyright The NOMAD Authors.
+#
+# This file is part of NOMAD. See https://nomad-lab.eu for further info.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from abc import ABC, abstractmethod
+from collections.abc import Generator
+from typing import Any, Callable, TypeVar, Union
+
+import numpy as np
+from nomad.metainfo.data_type import ExactNumber, InexactNumber
+
+T = TypeVar('T', bound='BoundedDataType')
+
+
+def _flatten_values(data: Any) -> Generator[Any, None, None]:
+    """Generator that yields all scalar values from nested list/array structure."""
+    if isinstance(data, np.ndarray):
+        yield from data.flatten()
+    elif isinstance(data, list):
+        for item in data:
+            if isinstance(item, list):
+                yield from _flatten_values(item)
+            else:
+                yield item
+    else:
+        yield data
+
+
+def _check_bounds(
+    value: Union[int, float],
```

Comment:
So obsolete, use `int | float`.
https://peps.python.org/pep-0604/

Reply:
Neat, I like this format way better! Is more in line with other type systems.
Since it is supported from 3.10 onward, I can apply it.